### PR TITLE
Add ResetOnCopy<T> and ReinitOnCopy<T> mixins

### PR DESCRIPTION
**Pull request by sherm1**
Add ResetOnCopy<T> and ReinitOnCopy<T> mixins. These are used on data members to provide resetting to default values on copy construction or copy assignment, or reinitialization to specified values.
These can be used on built-in types, pointers, containers, smart pointers, Simbody resource index objects, and almost anything else. (I don't know how to make them work with C array types though.)

Also:
- Added a clear() method to unique index types.
- Added templates that define all relational operators when given only operator==() and operator<().

### CODE REVIEW

**Comment by sherm1** [comment]
Some usage examples from the regression tests:

*Reset*OnCopy
```c++
// Member initializations here are "one time"; if this is copied the copy
// had all members default constructed instead. No overhead.
class UsesResetOnCopy {
public:
    ResetOnCopy<int>                     defint;  // gets zero-initialized
    ResetOnCopy<Body*>                   bodyptr; // gets nullptr-initialized
    ResetOnCopy<char>                    charZ = 'z';
    ResetOnCopy<string>                  defstr;  // gets default-initialized
    ResetOnCopy<string>                  strHelloC = "hello"; 
    ResetOnCopy<short>                   shArr[3] = {9,8,7};
    ResetOnCopy<SubsystemIndex>          subIx{5};
    ResetOnCopy<std::vector<string>>     vstr {"one", "two", "three"};
    ResetOnCopy<unique_ptr<Array_<int>>> up{new Array_<int>({1,2,3})}; // won't leak
};
```
*Reinit*OnCopy
```c++
// Member initializations here are remembered; copy construction using this as
// the source will result in the copy being restored to these values even if the source
// was changed. Overhead: an extra object of type T holds the initial value.
class UsesReinitOnCopy {
public:
    enum Color {Red, Green, Blue};

    // Default construction is not allowed for ReinitOnCopy.
    ReinitOnCopy<char>                   charZ = 'z';
    ReinitOnCopy<string>                 strHelloC = "hello";
    ReinitOnCopy<short>                  shArr[3] = {9,8,7};
    ReinitOnCopy<SubsystemIndex>         subIx{5};
    ReinitOnCopy<std::vector<string>>    vstr {"one", "two", "three"};
    ReinitOnCopy<Color>                  color{Blue};
};
```

For both classes move construction and assignment behaves normally. All methods are inherited from type T (including constructors and assignment operators) and there are default conversions to these members can be used as though they were T's rather than `ResetOnCopy<T>`'s.

**Comment by sherm1** [comment]
(please see earlier comments)
/cc @aseth1, @aymanhab, @klshrinidhi, @chrisdembia, @tkuchida, @msdemers, @thomasklau

**Comment by chrisdembia** [comment]
Cool stuff.

1. Do you plan to remove `NullOnCopyUniquePtr`?
2. What is the reason for different behavior between copy construction and copy assignment?
3. I was planning to use these classes for Simbody handles. The OpenSim Model dynamically allocates a few Simbody handles (MultibodySystem, SimbodyMatterSubsystem, Force::Gravity, ...) and I don't see a `reset()` or `clear()` on these classes; at least sometimes there is a `clearHandle()`. We can alternatively dynamically allocate the handles and use them with `ResetOnCopy`; just seems a little redundant though.

**Comment by sherm1** [comment]
> Do you plan to remove NullOnCopyUniquePtr?

Yes. These can also be used instead of ReferencePtr, but I am not sure which is better.

> What is the reason for different behavior between copy construction and copy assignment?

I'm not sure what you mean. They both perform reset or reinit. Please clarify.

> I was planning to use these classes for Simbody handles. 

I'm planning to add clear() methods to all handles if these classes look promising.

**Comment by chrisdembia** [comment]
> I'm not sure what you mean. They both perform reset or reinit. Please clarify.

The doxygen comment fro `ResetOnCopy` says:
```
For class types `T`, copy construction will use the 
default constructor, and copy assignment will invoke `T::reset()` or, if there 
is no such method then `T::clear()`.
```

Isn't this different behavior? The copy assignment could also use the default constructor.

**Comment by sherm1** [comment]
The test failure (only on OSX) appears to be a bug in the Apple std::string library's implementation of move assignment. This program demonstrates the problem:
```c++
#include <string>
#include <iostream>
#include <utility>
using std::cout; using std::endl;

int main() {
    std::string s2("goodbye");
    cout << "s2=" << s2 << endl;

    s2 = s2;
    cout << "s2=s2 -> s2=" << s2 << endl;
    s2 = std::move(s2);
    cout << "s2=move(s2) -> s2=" << s2 << endl;
}
```
gcc, Linux clang, and VS2015 agree that the self-move shouldn't change the string's value. On OSX only (at least for clang 3.5 there), the string gets wiped out by the self move. I'll have to fix the test for now; does anyone have a later OSX compiler to check if this has been fixed?

**Comment by chrisdembia** [comment]
> Apple std::string library's implementation of move assignment.

Perhaps it's a difference between libstdc++ and LLVM's libc++.

**Comment by chrisdembia** [comment]
Output on clang 3.6 (Apple LLVM version 6.1.0) on OSX 10.10.3:

```
s2=goodbye
s2=s2 -> s2=goodbye
s2=move(s2) -> s2=
```

**Comment by chrisdembia** [comment]
On Ubuntu, using clang 3.6 and libc++ instead of libstdc++, the bug still exists. So I think it's libc++.

**Comment by sherm1** [comment]
> Isn't this different behavior? The copy assignment could also use the default constructor.

Ah, I see. It can't use the default constructor for `this` because the object already exists. If it default constructs a value it would have to copy assign that value to `this`, but `ResetOnCopy<T>` can't assume that T has a copy assignment operator (std::unique_ptr, for example, does not). 

But now that you've said that, I'm wondering if it would be reasonable to invoke the object's destructor and then use "placement new" to default construct it again in place?

**Comment by sherm1** [comment]
> On Ubuntu, using clang 3.6 and libc++ instead of libstdc++, the bug still exists. So I think it's libc++.

Thanks. I wonder how to file a bug report for that.

**Comment by chrisdembia** [comment]
Hmm...maybe [here](https://llvm.org/bugs/)?

**Comment by klshrinidhi** [comment]
I looked into the question:
**Is it allowed and safe to explicitly call the destructor of an object and default construct the object in its place ?**
Short answer -- Yes.
Long answer -- Like I mentioned in the meeting today, I think the problem is more with objects created on the stack and not so much with objects created on the heap because we have control over lifetime of heap objects. I wanted to check if calling the destructor explicitly releases memory for stack objects, and looks like it doesn't. For heap objects, `delete` operator does two things -- destroy and deallocate memory, and so calling destructor explicitly for heap objects will do only that. For stack objects, compiler inserts code to destroy and deallocate when objects go out of scope. If there is *some* valid object in that location when end of scope is reached, there should be no problem.

Note:- I have not tested this with real code.

**Comment by sherm1** [comment]
> I looked into the question: Is it allowed and safe to explicitly call the destructor of an object and default construct the object in its place?

Thanks, Shrinidhi! That's encouraging. Should work the same for heap and stack objects because the code executing in the copy assignment operator will be the same code. I'm going to try this and see if I can pass all my test cases.

**Comment by sherm1** [comment]
> Hmm...maybe here?

Thanks. Filed LLVM bugreport here: https://llvm.org/bugs/show_bug.cgi?id=24453

I switched to `std::vector` from `std::string` for my test case and found it failed on gcc and clang everywhere. So I switched to `SimTK::Array_` where at least I know the move constructor works right!

**Comment by sherm1** [comment]
I tried the Shrinidhi-approved destruct/construct approach in ResetOnCopy<T> and it seems to work fine on all platforms! Here is what copy assignment looks like now:
```c++
    // Here T is of unknown class type; built-ins are handled more easily elsewhere.
    ResetOnCopyHelper& operator=(const ResetOnCopyHelper&) {
        T* thisT = static_cast<T*>(this); // upcast to base
        thisT->~T();        // destruct the current base object
        new (thisT) T();    // reconstruct it with default constructor
        return *this;
    }
```
So now this class should work on anything T that has a destructor and default constructor -- no need for `reset()` or `clear()` methods. Very cool!

**Comment by sherm1** [comment]
This PR is ready for review if any of you brave souls would like to take a crack at it.

**Comment by chrisdembia** [comment]
> thisT->~T();

Are you sure this won't call `transpose()` when `T = SimTK::Matrix`, etc.?

**Comment by sherm1** [comment]
> Are you sure this won't call transpose() when T = SimTK::Matrix, etc.?

Good question, but I think this is safe. The method you are thinking of is actually named `T::operator~()` not `T::~T()`.

**Comment by sherm1** [comment]
The clang people are very responsive! I already got lots of comments, explaining that the standard library is intentionally noncommittal about self-move behavior:

Comment # 2 on bug 24453 from Marshall Clow (home) 
In general there is no requirement on self-move assignment for the standard library. In particular, there is no requirement that it be a no-op.

Howard Hinnant has written a good explanation here:
http://stackoverflow.com/questions/13127455/what-does-the-standard-library-guarantee-about-self-move-assignment

**Comment by chrisdembia** [comment]
> Good question, but I think this is safe. The method you are thinking of is actually named T::operator~() not T::~T().

Oh okay. I guess `this->~T()` is different from `~this->T()`.

**Comment by sherm1** [comment]
> Oh okay. I guess this->~T() is different from ~this->T().

Yes, but I'm adding some more tests to prove it.

File: SimTKcommon/tests/TestCloneOnWritePtr.cpp [hunk 1]
```diff
@@ -426,6 +432,287 @@ void testNullOnCopyUniquePtr() {
...
+    // self copy of underlying std::vector should do nothing
+    myvec.updT() = myvec.getT();
+    SimTK_TEST(myvec.size()==3 && myvec[2]==3 && &myvec[2]==addr2);
+
+    // But, self-copy assignment should clear (cuz reset-on-copy)
+    myvec = myvec;
+    SimTK_TEST(myvec.empty());
+
+    // Test with a matrix.
+    ResetOnCopy<Mat<2,3>> mat23(1,2,3,
```
Comment (sherm1) [inline SimTKcommon/tests/TestCloneOnWritePtr.cpp]: Matrix tests start here, @chrisdembia.

**Comment by sherm1** [comment]
Did some cleanup and doxygen tweaks. I'm done with these now. Any further comments before I merge this? I haven't removed NullOnCopyUniquePtr but will when it isn't being used any more. (`ResetOnCopy<unique_ptr<T>>` replaces it.)

**Comment by chrisdembia** [comment]
No further comments from me.

**Comment by chrisdembia** [comment]
Might be nice to have an alias `using NullOnCopy<T> = ResetOnCopy<unique_ptr<T>>`.

**Comment by sherm1** [comment]
> Might be nice to have an alias using NullOnCopy<T> = ResetOnCopy<unique_ptr<T>>.

I'm reluctant to hide the object ownership consideration. Having it visible could avoid trouble later. Hopefully we'll get used to thinking:
```c++
    ResetOnCopy<double*>           m_thing; // <<<--- RED FLAG, possible leak
    ResetOnCopy<unique_ptr<double> m_thing; // all good here
```

File: SimTKcommon/include/SimTKcommon/internal/ResetOnCopy.h [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    const T& getT() const {return m_value;}
+    T&       updT()       {return m_value;}
+
+private:
+    T   m_value{}; // value initialize; i.e., set to zero
+};
+
+
+/** ResetOnCopy helper class specialization for any type `T` that is not a
+built-in ("scalar") type and that has a reset() or clear() method. This will
```
Comment (chrisdembia) [inline SimTKcommon/include/SimTKcommon/internal/ResetOnCopy.h]: I don't think this comment is correct anymore.

Comment (sherm1) [inline SimTKcommon/include/SimTKcommon/internal/ResetOnCopy.h]: Thanks Chris. I added a fix to pending PR #427.

**Pull request merged**